src/Workspaces/Remote/ServiceHub/Services/CustomMessageHandler/RemoteCustomMessageHandlerService.cs

```diff
@@ -0,0 +1,252 @@
+                            }
+
+                            return sharedAssembly;
+                        }
+
+                        var extensionAssemblyPath = Path.Combine(directory, $"{assemblyName.Name}.dll");
+                        if (File.Exists(extensionAssemblyPath))
+                        {
+                            return loadContext.LoadFromAssemblyPath(extensionAssemblyPath);
+                        }
+
+                        return null;
+                    };
+
+                    return new CustomMessageHandlerExtension(loadContext);
+                });
+
+                var handler = extension.Handlers.GetOrAdd(typeFullName, name =>
+                {
+                    var assembly = extension.AssemblyLoadContext.LoadFromAssemblyPath(assemblyPath);
+                    var type = assembly.GetType(typeFullName)
+                        ?? throw new InvalidOperationException($"Cannot find type {typeFullName} in {assemblyPath}.");
+
+                    // Create the Handler instance. Requires having a parameterless constructor.
+                    // ```
+                    // public class CustomMessageHandler
+                    // {
+                    //     public Task<TResponse> ExecuteAsync(TRequest, CancellationToken);
+                    // }
+                    //
+                    // public class CustomMessageDocumentHandler
+                    // {
+                    //     public Task<TResponse> ExecuteAsync(TRequest, Document, CancellationToken);
+                    // }
+                    // ```
+                    return Activator.CreateInstance(type)
+                        ?? throw new InvalidOperationException($"Cannot create {typeFullName} from {assemblyPath}.");
+                });
+
+                // TODO: use a well-known interface once available
+                const string executeMethodName = "ExecuteAsync";
+                var executeMethod = handler.GetType().GetMethod(executeMethodName, BindingFlags.Public | BindingFlags.Instance)
+                    ?? throw new InvalidOperationException($"Cannot find method {executeMethodName} in type {typeFullName} assembly {assemblyPath}.");
+
+                // CustomMessage.Message references positions in CustomMessage.TextDocument as indexes referencing CustomMessage.Positions.
+                // LinePositionReadConverter allows the deserialization of these indexes into LinePosition objects.
+                JsonSerializerOptions readOptions = new();
+                if (document is not null)
+                {
+                    LinePositionReadConverter linePositionReadConverter = new(positions);
+                    readOptions.Converters.Add(linePositionReadConverter);
+                }
+
+                // Deserialize the message into the expected TRequest type.
+                var requestType = executeMethod.GetParameters()[0].ParameterType;
```

Comment:
the method info and parameter types should likely be cached inside the CustomMessageHandlerExtension